### DISCUSSION TITLE
RSA_PRIVATE_KEY error

### DISCUSSION THREAD

**Discussion by usama-mashood** [saleor/saleor#12226] [Deployment]
**i am getting this error when i set debug=False and set RSA_PRIVATE_KEY in production**

2023-03-06T06:25:03.899956181Z File "/app/saleor/core/jwt_manager.py", line 200, in validate_configuration
2023-03-06T06:25:03.899962001Z raise ImproperlyConfigured(f"Unable to load provided PEM private key. {e}")
2023-03-06T06:25:03.899967751Z django.core.exceptions.ImproperlyConfigured: Unable to load provided PEM private key. ('Could not deserialize key data. The data may be in an incorrect format, it may be encrypted with an unsupported algorithm, or it may be an unsupported key type (e.g. EC curves with explicit parameters).', [_OpenSSLErrorWithText(code=503841036, lib=60, reason=524556, reason_text=b'error:1E08010C:DECODER routines::unsupported')])

**Comment by usama-mashood** [comment]
**I have followed all the solutions suggested in the link but not getting any luck**
https://github.com/saleor/saleor/discussions/9331

**Comment by korycins** [comment]
The `RSA_PRIVATE_KEY` variable should contain the content of your private key in PEM format. You can generate the key and set the content of the file inside the ENV variable using one of the commands below.

```
ssh-keygen -m PEM -t rsa -b 2048 -f private.pem
export RSA_PRIVATE_KEY=`cat private.pem` << this is only to show that you need to copy your private.pem key to Saleor's variable>>
export RSA_PRIVATE_PASSWORD="<<the password to pem file>>"
```
or
```
openssl genrsa -out private-key.pem 2048
export RSA_PRIVATE_KEY=`cat private-key.pem` << this is only to show that you need to copy your private.pem key to Saleor's variable>>
```

**Reply by usama-mashood** [reply]
I followed these steps but still got the same error

**Comment by paulrollo** [comment]
I've got saleor core in a docker container. When I'm switching from dev to production I change DEBUG to 'False' and i've got issues with my RSA_PRIVATE_KEY.  I've read every document on the web which includes RSA_PRIVATE_KEY and saleor. Still, I'm stuck with a container that doesn't go up and this in the log;
```
File "/app/saleor/core/jwt_manager.py", line 200, in validate_configuration
    raise ImproperlyConfigured(f"Unable to load provided PEM private key. {e}")
django.core.exceptions.ImproperlyConfigured: Unable to load provided PEM private key. ('Could not deserialize key data. The data may be in an incorrect format, it may be encrypted with an unsupported algorithm, or it may be an unsupported key type (e.g. EC curves with explicit parameters).', [_OpenSSLErrorWithText(code=503841036, lib=60, reason=524556, reason_text=b'error:1E08010C:DECODER routines::unsupported')]) 
```

I've followed @korycins note above to generate the key with
`ssh-keygen -m PEM -t rsa -b 2048 -f private.pem`

My docker-compose.yaml is

```
version: '3'
services:
  web:
    image: ghcr.io/saleor/saleor
    ports:
      - "127.0.0.1:8000:8000"
    extra_hosts:
      - host.docker.internal:host-gateway
    environment:
      ALLOWED_HOSTS: '*'
      ALLOWED_CLIENT_HOSTS: '*'
      ALLOWED_GRAPHQL_ORIGINS: '*'
      DEBUG: 'False'
      DATABASE_URL: postgres://saleor:replaced-password@host.docker.internal:5432/saleor
      SECRET_KEY: 'replaced-secret-key'
      REDIS_URL: redis://host.docker.internal:6379
      LOAD_EXAMPLE_DATA: 'True'
      CELERY_BROKER_URL: redis://host.docker.internal:6379/1
      DEFAULT_FROM_EMAIL: support@example.com
      DEFAULT_COUNTRY: 'GB'
      DEFAULT_CURRENCY: 'GBP'
      RSA_PRIVATE_KEY: /run/secrets/private_key
      RSA_PRIVATE_PASSWORD: 'password'
    secrets:
      - private_key
    volumes:
      - ./media:/app/media

secrets:
   private_key:
     file: private.pem
```
I've tried stripping newlines, replacing newlines with \n, including the RSA_PRIVATE_KEY inline in the docker-compose.yaml. I've tried generating keys a hundred different ways. The permissions on my private.pem are 
```
-rw------- 1 saleor saleor 1766 Apr 30 00:01 private.pem
```
but, I've tried 777 just to rule that out. 

It seems to be a problem that has come up a bunch of times, and feels like it should be straight forward.  I would love an authoritative answer on how to do this with docker compose.  Anyone help would be massively appreciated.

**Reply by korycins** [reply]
Hey
Just tried to test it. I have a docker-compose like below
```
version: "3.4"

services:
  api:
    image: ghcr.io/saleor/saleor:3.12.0
    ports:
      - 8000:8000
    restart: unless-stopped
    networks:
      - saleor-backend-tier
    stdin_open: true
    tty: true
    depends_on:
      - db
      - redis
      - jaeger
    volumes:
      # shared volume between worker and api for media
      - saleor-media:/app/media
    env_file:
      - common.env
      - backend.env
    environment:
      - ALLOWED_CLIENT_HOSTS
      - RSA_PRIVATE_KEY
      - RSA_PRIVATE_PASSWORD
      - DEBUG=False
      - JAEGER_AGENT_HOST=jaeger
      - DASHBOARD_URL=http://localhost:9000/
      - ALLOWED_HOSTS=localhost,api
```
Then I am calling
`ssh-keygen -m PEM -t rsa -b 2048 -f private.pem`
set password as `password`

Then:
```
 export RSA_PRIVATE_KEY=`cat private.pem`
 export RSA_PRIVATE_PASSWORD="password"
 export DEBUG=False
 export ALLOWED_CLIENT_HOSTS=localhost,127.0.0.1,localhost:3000,localhost:9000,localhost:4173
```

Then docker-compose runs `api` without a problem:
```
➜  saleor-platform git:(main) ✗ docker-compose up api
saleor-platform_jaeger_1 is up-to-date
saleor-platform_redis_1 is up-to-date
saleor-platform_db_1 is up-to-date
Recreating saleor-platform_api_1 ... done
Attaching to saleor-platform_api_1
api_1        | [2023-05-09 07:22:16 +0000] [1] [INFO] Starting gunicorn 20.1.0
api_1        | [2023-05-09 07:22:16 +0000] [1] [INFO] Listening at: http://0.0.0.0:8000 (1)
api_1        | [2023-05-09 07:22:16 +0000] [1] [INFO] Using worker: saleor.asgi.gunicorn_worker.UvicornWorker
api_1        | [2023-05-09 07:22:16 +0000] [23] [INFO] Booting worker with pid: 23
api_1        | [2023-05-09 07:22:16 +0000] [23] [INFO] Started server process [23]
api_1        | [2023-05-09 07:22:16 +0000] [24] [INFO] Booting worker with pid: 24
api_1        | [2023-05-09 07:22:16 +0000] [24] [INFO] Started server process [24]
api_1        | [2023-05-09 07:22:16 +0000] [25] [INFO] Booting worker with pid: 25
api_1        | [2023-05-09 07:22:16 +0000] [26] [INFO] Booting worker with pid: 26
api_1        | [2023-05-09 07:22:16 +0000] [25] [INFO] Started server process [25]
api_1        | [2023-05-09 07:22:17 +0000] [26] [INFO] Started server process [26]
```

Hope that this will help you 🤞
